### ISSUE TITLE
Subcomponents are not effective

### ISSUE THREAD

**Issue by MQYForverT** [unocss/unocss#3920]
### UnoCSS version

0.61.0

### Describe the bug

I use Vue3 to write web component, now i have two component. A & B

I imported B  in A (B is not custom elements), and then used A as a custom element for external use.
I used a placeholder in A, which is available, but unocss in B is invalid. I tried to add a placeholder in B as well, but it was also invalid

After my testing, I found that the style needs to be declared once in component A for it to take effect in component B

So if there are many sub components in my lower layer, and all the styles they use are defined meaninglessly once in the root component, doesn't that feel strange?

I think this is a bug that has seriously affected the use of a series of composite web components and is very uncomfortable

### Reproduction

https://stackblitz.com/edit/vitejs-vite-d8hur1?file=src%2Fcomponents%2Findex.ts

### System Info

node：20.11.0
pnpm: 9.0.6

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/unocss/unocss/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't [already an issue](https://github.com/unocss/unocss/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/unocss/unocss/discussions) or join our [Discord Chat Server](https://chat.antfu.me).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by Simon-He95** [comment]
We temporarily closed this due to the lack of enough information. We could not identify whether it was a bug or a userland misconfiguration with the given info.
Please provide a [minimal reproduction](https://stackoverflow.com/help/minimal-reproducible-example) to reopen the issue.
Thanks.

> [Why reproduction is required](https://antfu.me/posts/why-reproductions-are-required)

**Comment by MQYForverT** [comment]
> 由于缺乏足够的信息，我们暂时关闭了此问题。根据给出的信息，我们无法确定这是错误还是用户空间配置错误。 请提供[最小限度的重现](https://stackoverflow.com/help/minimal-reproducible-example)以重新打开此问题。 谢谢。
> 
> > [为什么需要繁殖](https://antfu.me/posts/why-reproductions-are-required)

https://stackblitz.com/edit/vitejs-vite-d8hur1?file=src%2Fcomponents%2Findex.ts

**Comment by MQYForverT** [comment]
@Simon-He95 May I ask if the problem has been identified and if there are any repair arrangements in place

**Comment by Simon-He95** [comment]
> @Simon-He95 May I ask if the problem has been identified and if there are any repair arrangements in place请问是否已发现问题，以及是否有任何维修安排

Sorry, I have never used this form. I need to look at the implementation first.

**Comment by userquin** [comment]
Shadow dom styles needs a deep review. The problem here is about styles in the child B component and the vue hmr styles, the css is there but  seems vue cannot apply them (no idea why yet 🤔 ).

You can add vite-plugin-inspect to your sb repro, and check virtual modules checkbox, you will see both components and the corresponding inlined css styles.

I also tried adding a second global UnoCSS plugin, the styles are there (duplicated) but still not working.

**Comment by MQYForverT** [comment]
> Shadow dom 样式需要深入研究。这里的问题是关于子 B 组件中的样式和 vue hmr 样式，css 在那里，但 vue 似乎无法应用它们（目前还不知道为什么 🤔 ）。
> 
> 您可以将 vite-plugin-inspect 添加到您的 sb repro，并选中虚拟模块并设置其相应的内部联 css 样式。
> 
> 我尝试添加第二个全局 UnoCSS 插件，样式在那里（重复）但仍然不起作用。

Thank you for your investigation. Do we currently have any temporary solutions before fixing the problem? Can we only define all the styles used once in the root component?

**Comment by userquin** [comment]
The problem is the shadow dom, styles doesn't work as normal styles since they are scoped, slotted and so on (similar to deep in vue). We also have the :root concept and the styles are not applied properly. Check the issues, there are a few entries, also in discord channel (uhmm, iirc in unocss team channel, I will send to unocss some questions @Simon-He95 or maybe here)

Since we have now generators, I Will check if we can improve them.

**Comment by Simon-He95** [comment]
[defineCustomElement](https://cn.vuejs.org/guide/extras/web-components.html#sfc-as-custom-element)
There seems to be no such component for sfc to import components, but it can be handled in the following way
<img width="799" alt="image" src="https://github.com/unocss/unocss/assets/57086651/c9fa616a-cef3-4ed4-9830-a1a7cd7c6c56">

If you manually merge the styles of the dependent components into the style of the root component, you can achieve your expectations.
This should solve your current problem. @MQYForverT

**Comment by userquin** [comment]
will this break the shadow dom scope? can you show the styles in CompA?

**Comment by Simon-He95** [comment]
> will this break the shadow dom scope? can you show the styles in CompA?

<img width="1384" alt="image" src="https://github.com/unocss/unocss/assets/57086651/6408a666-86c6-4375-8de3-7e1bc821dd5a">

**Comment by userquin** [comment]
so, the CompB is just rendered, it is not a web component, weird use case, this can break other child components styles since it should use [::slotted](https://developer.mozilla.org/en-US/docs/Web/CSS/::slotted)

**Comment by MQYForverT** [comment]
<img width="1356" alt="image" src="https://github.com/unocss/unocss/assets/36668907/85e9264a-5af4-4ac6-a4ba-4b06751883d6">
CompB is not a web component, which means they are a whole, and this whole is in a shadowdom. I found that the unocss of compB are not generated in the style tag

**Comment by MQYForverT** [comment]
> [defineCustomElement](https://cn.vuejs.org/guide/extras/web-components.html#sfc-as-custom-element) sfc 导入组件好像没有这个组件，不过可以按照下面的方法处理 <img alt="图像" width="799" src="https://private-user-images.githubusercontent.com/57086651/343695479-c9fa616a-cef3-4ed4-9830-a1a7cd7c6c56.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0NzgzMjksIm5iZiI6MTcxOTQ3ODAyOSwicGF0aCI6Ii81NzA4NjY1MS8zNDM2OTU0NzktYzlmYTYxNmEtY2VmMy00ZWQ0LTk4MzAtYTFhN2NkN2M2YzU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI3VDA4NDcwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0Y2NlN2M1NGNmN2Y5YTA3NWI2YWE5NDc3YjFjYzNkNjQ4NmQ1Y2ZiMTZkZDc4ZmQxYTQ5MzM4ZmZkNDMyMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._RaqfR_kQg7lpETBAW3NFhvSEpqO96uBhGPmrpJiDqU">
> 
> 如果你手动将依赖组件的样式合并到根组件的样式中，就可以达到你的预期。 这应该可以解决你目前的问题。@MQYForverT

A temporary solution requires two steps
1. When using unocss components, @ unocss placeholder needs to be introduced;
2. Introducing styles for all subcomponents in the root component may solve the problem, but it doesn't feel very elegant and can be even more uncomfortable when there are many subcomponents
I think the most reasonable thing is that all the unocss of the child components should appear in the style tag of the root component

**Comment by Simon-He95** [comment]
> so, the CompB is just rendered, it is not a web component, weird use case, this can break other child components styles since it should use [::slotted](https://developer.mozilla.org/en-US/docs/Web/CSS/::slotted)

This scenario is not limited to unocss. The normal css written in the component will not take effect. A better way is to let vue improve defineCustomElement.

**Comment by userquin** [comment]
yeah, but them we will need to traverse the full tree per component (checking if the target is a web component, no idea how to check this), that's not affordable... (you can have a web component using vue component using web component using vue component...)

**Comment by userquin** [comment]
@Simon-He95 maybe we have somthing wrong in the plugin, check https://github.com/unocss/unocss/blob/main/packages/vite/src/modes/shadow-dom.ts#L103-L108

**Comment by MQYForverT** [comment]
> > 因此，CompB 只是被渲染了，它不是一个 Web 组件，这是一个奇怪的用例，因为它应该使用[::slotted，所以这可能会破坏其他子组件的样式](https://developer.mozilla.org/en-US/docs/Web/CSS/::slotted)
> 
> 这种场景不仅限于unocss，组件里写的普通css不会生效，更好的办法是让vue完善defineCustomElement。

I will use a temporary solution first, and I will also initiate a discussion on Vue's side. We can discuss together how to perfectly solve this problem

**Comment by hickiy** [comment]
I think this is a normally problem,  comB isn't to been resolve by unocss plugin,  so its class haven't to been transifer and insert,  so that is a compatbility problem between unocss and vite
